### PR TITLE
Vm Idler

### PR DESCRIPTION
**Pull request by rajivnathan**
https://issues.redhat.com/browse/SANDBOX-376

No e2e tests for idling VMs due to infra constraints.

### CODE REVIEW

File: test/idler_assertion.go [hunk 1]
```diff
@@ -266,3 +273,23 @@ func (a *IdleablePayloadAssertion) StatefulSetScaledUp(statefulSet *appsv1.State
...
+
+func (a *IdleablePayloadAssertion) VMStopped(vm *unstructured.Unstructured) *IdleablePayloadAssertion {
+	vm, err := a.dynamicClient.Resource(vmGVR).Namespace(vm.GetNamespace()).Get(context.TODO(), vm.GetName(), metav1.GetOptions{})
+	require.NoError(a.t, err)
+	val, found, err := unstructured.NestedBool(vm.Object, "spec", "running")
+	require.NoError(a.t, err)
+	assert.True(a.t, found)
+	assert.False(a.t, val)
+	return a
+}
```
Comment (alexeykazakov) [inline test/idler_assertion.go:295]: Or something like this:

```suggestion
func (a *IdleablePayloadAssertion) VMRunning(vm *unstructured.Unstructured) *IdleablePayloadAssertion {
	return a.vmRunning(vm, true)
}

func (a *IdleablePayloadAssertion) VMStopped(vm *unstructured.Unstructured) *IdleablePayloadAssertion {
	return a.vmRunning(vm, false)
}

func (a *IdleablePayloadAssertion) vmRunning(vm *unstructured.Unstructured, running bool) *IdleablePayloadAssertion {
	vm, err := a.dynamicClient.Resource(vmGVR).Namespace(vm.GetNamespace()).Get(context.TODO(), vm.GetName(), metav1.GetOptions{})
	require.NoError(a.t, err)
	val, found, err := unstructured.NestedBool(vm.Object, "spec", "running")
	require.NoError(a.t, err)
	assert.True(a.t, found)
	assert.Equal(a.t, running, val)
	return a
}
```

File: controllers/idler/idler_controller_test.go [hunk 2]
```diff
@@ -518,12 +579,14 @@ func TestEnsureIdlingFailed(t *testing.T) {
 			assertCanNotUpdateObject(&appsv1.StatefulSet{}, "can't update statefulset")
 			assertCanNotUpdateObject(&openshiftappsv1.DeploymentConfig{}, "can't update deploymentconfig")
 			assertCanNotUpdateObject(&corev1.ReplicationController{}, "can't update replicationcontroller")
+			assertCanNotUpdateObject(&corev1.ReplicationController{}, "can't update replicationcontroller")
```
Comment (alexeykazakov) [inline controllers/idler/idler_controller_test.go:582]: A duplicate by mistake?
```suggestion
```

**Review by alexeykazakov — APPROVED** [review]
Nice! Just a couple of minor comments.

File: controllers/idler/idler_controller_test.go (hunk 2 above)
Comment (rajivnathan) [reply controllers/idler/idler_controller_test.go:582]: Good catch!

File: test/idler_assertion.go (hunk 1 above)
Comment (rajivnathan) [reply test/idler_assertion.go:295]: Yeah, I actually did initially have a function `vmRunning(vm *unstructured.Unstructured, running bool)` but liked `VMRunning` and `VMStopped` from the test perspective. This is a good compromise. 👍

**Comment by alexeykazakov** [comment]
Let's see if https://github.com/codeready-toolchain/toolchain-e2e/pull/856 fixes the failed test.

**Pull request merged**